From d61574e6918a05891667c01633bec711de570bae Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Tomasz=20Paku=C5=82a?= <tomasz.pakula.oficjalny@gmail.com>
Date: Wed, 14 Jan 2026 20:28:06 +0100
Subject: [PATCH 10/11] drm/amd/display: try supporting HDMI VRRmax=0
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

VRRmax=0 is a valid value and means that the upper bound is guared by
the selected video mode. Try getting upper VRR bound from AMD vsdb
if it exists.

Signed-off-by: Tomasz Pakuła <tomasz.pakula.oficjalny@gmail.com>
---
 drivers/gpu/drm/amd/display/amdgpu_dm/amdgpu_dm.c | 9 +++++++--
 1 file changed, 7 insertions(+), 2 deletions(-)

diff --git a/drivers/gpu/drm/amd/display/amdgpu_dm/amdgpu_dm.c b/drivers/gpu/drm/amd/display/amdgpu_dm/amdgpu_dm.c
index 79f105a96e89..9ac0441d393b 100644
--- a/drivers/gpu/drm/amd/display/amdgpu_dm/amdgpu_dm.c
+++ b/drivers/gpu/drm/amd/display/amdgpu_dm/amdgpu_dm.c
@@ -13066,12 +13066,17 @@ void amdgpu_dm_update_freesync_caps(struct drm_connector *connector,
 
 	if (as_type == ADAPTIVE_SYNC_TYPE_PCON_IN_WHITELIST) {
 		/* Prefer HDMI VRR over FreeSync */
-		if (hdmi_vrr->supported && hdmi_vrr->vrr_max > 0) {
+		if (hdmi_vrr->supported) {
 			amdgpu_dm_connector->pack_sdp_v1_3 = true;
 			amdgpu_dm_connector->as_type = as_type;
 
+			/* 0 is a valid VRRmax. Fetch upper bound from AMD vsdb if present */
+			if (hdmi_vrr->vrr_max == 0 && vsdb_freesync)
+				monitor_range_from_vsdb(connector, &vsdb_info);
+			else
+				connector->display_info.monitor_range.max_vfreq = hdmi_vrr->vrr_max;
+
 			connector->display_info.monitor_range.min_vfreq = hdmi_vrr->vrr_min;
-			connector->display_info.monitor_range.max_vfreq = hdmi_vrr->vrr_max;
 			freesync_capable = copy_range_to_amdgpu_connector(connector);
 
 		} else if (vsdb_freesync) {
-- 
2.52.0

